src/bio/bio_context.c

```diff
@@ -599,6 +599,28 @@ __bio_ioctxt_open(struct bio_io_context **pctxt, struct bio_xs_context *xs_ctxt,
 	ctxt->bic_xs_ctxt = xs_ctxt;
 	uuid_copy(ctxt->bic_pool_id, uuid);
 
+	/* init WAL replay metrics */
+	if (st == SMD_DEV_TYPE_WAL) {
+		struct bio_dma_stats	*stats = &xs_ctxt->bxc_dma_buf->bdb_stats;
+		char			 pool[DAOS_UUID_STR_SIZE];
+
+		uuid_unparse_lower(uuid, pool);
+		rc = d_tm_add_metric(&stats->bds_wal_rpl_sz, D_TM_GAUGE,
+				     "WAL replay size", "bytes",
+				     "dmabuff/wal_replay/tgt_%d/pool/%s/size",
+				     xs_ctxt->bxc_tgt_id, pool);
+		if (rc)
+			D_WARN("Failed to create WAL size replay telemetry: "DF_RC"\n", DP_RC(rc));
+
+		rc = d_tm_add_metric(&stats->bds_wal_rpl_tm,
+				     D_TM_DURATION | D_TM_CLOCK_THREAD_CPUTIME,
+				     "WAL replay time", "us",
+				     "dmabuff/wal_replay/tgt_%d/pool/%s/time",
+				     xs_ctxt->bxc_tgt_id, pool);
+		if (rc)
+			D_WARN("Failed to create WAL replay ET telemetry: "DF_RC"\n", DP_RC(rc));
+	}
```

Comment:
1a. d_tm_add_metric() function adds a metric by path. If the path exists, the existing metric would be re-used, plz see src/gurt/telemetry.c::d_tm_add_metric() lines 2256..2261.
So we do create the metrics by pool, yes.
If path doesn't exist, add_metric() would create a new metric node in shmem.
The allocations get free at d_tm_del_ephemeral_dir() by the metric root which is "dmabuff/" in this case.
So there is no memory leaks no matter which and how many pools are open. I can confirm that with valgrind --memcheck

1b. It is not a big deal to make the metric per-pool: that would require just the path format modification in lines 
"dmabuff/wal_replay/tgt_%d/pool/%s/time" ! We could do that any time in future.

2. We could move the initialization calls to dma_metrics_init() making the metrics per-engine, yes. I see no benefit in that because the current per-pool metric is more versatile.

Reply:
I mean now we have metrics grouped by engine (tgt_id) first, then by pool uuids.

Reply:
Ok, then let's make it per-pool metric. Could you implement in the same way we used for other server modules? (see dss_module_metrcis, ds_pool_metrcis_start() and ds_pool_metrics_stop())

Reply:
Please see the updated patch: it implements both per-engine and per-pool metrics.